packages/upload-client/src/fetch-with-upload-progress.js

```diff
@@ -1,8 +1,170 @@
-import ipfsUtilsFetch from 'ipfs-utils/src/http/fetch.js'
+// HTTP/2 or Node.js/undici
+
+/**
+ *
+ * @param {AsyncIterable<Uint8Array>} iterable
+ * @returns {ReadableStream}
+ */
+function iterableToStream(iterable) {
+  const iterator = iterable[Symbol.asyncIterator]()
+  return new ReadableStream({
+    async pull(controller) {
+      const { value, done } = await iterator.next()
+      if (value) {
+        controller.enqueue(value)
+      }
+      if (done) {
+        controller.close()
+      }
+    },
+  })
+}
+
+/**
+ * Takes body from fetch response as body and `onUploadProgress` handler
+ * and returns async iterable that emits body chunks and emits
+ * `onUploadProgress`.
+ *
+ * @param {ReadableStream} body
+ * @param {import('./types.js').ProgressFn} onUploadProgress
+ * @returns {AsyncIterable<Uint8Array>}
+ */
+export const iterateBodyWithProgress = async function* (
+  body,
+  onUploadProgress
+) {
+  if (body instanceof ReadableStream) {
+    const reader = body.getReader()
+    const total = 0 // If the total size is unknown
+    const lengthComputable = false
+    let loaded = 0
+
+    try {
+      while (true) {
+        const { done, value } = await reader.read()
+        if (done) break
+
+        loaded += value.byteLength
+        yield value // Yield the chunk
+        onUploadProgress({ total, loaded, lengthComputable })
+      }
+      // eslint-disable-next-line no-useless-catch
+    } catch (e) {
+      throw e
+    } finally {
+      reader.releaseLock() // Ensure the reader lock is released
+    }
+  } else {
+    throw new Error('Body is not a ReadableStream')
+  }
+}
+
+/**
+ * Takes fetch options and wraps request body to track upload progress if
+ * `onUploadProgress` is supplied. Otherwise returns options as is.
+ *
+ * @param {import('./types.js').FetchOptions} options
+ * @returns {import('./types.js').FetchOptions}
+ */
+const withUploadProgress = (options) => {
+  const { onUploadProgress, body } = options
+
+  if (!onUploadProgress) {
+    return options
+  }
+
+  const rsp = new Response(body)
+  // @ts-expect-error web streams from node and web have different types
+  const source = iterateBodyWithProgress(rsp.body, onUploadProgress)
```

Comment:
Why not pass `body` directly to `iterateBodyWithProgress`? Can you add a comment if this is necessary or change to pass directly? 🙏